src/v/storage/disk_log_impl.cc

```diff
@@ -1182,25 +1190,21 @@ ss::future<>
 disk_log_impl::update_configuration(ntp_config::default_overrides o) {
     auto was_compacted = config().is_compacted();
     mutable_config().set_overrides(o);
+
     /**
      * For most of the settings we always query ntp config, only cleanup_policy
-     * and segment size need special treatment.
+     * needs special treatment.
```

Comment:
nit: would be nice to leave a comment explaining the limitations of this method so future developers don't make the same mistake.

Are we missing a hook in the topic config update path that would make this work as expected?

Reply:
I added a comment here.

> Are we missing a hook in the topic config update path that would make this work as expected?

So the thing is there are two different ways to deal with cluster level and topic level config updates For the first one, we have to bind to a config property of interest whereas for the latter we have hooks like these ^^. Underlying reason is that we have two different state machines that propagate these updates and they call into different places. Unfortunately unifying them into a single place doesn't seem trivial given the way the code is structured (Needs a lot of cleanup in log_config and ntp_config). Now to make things more confusing, we have some configs that only have topic level overrides but no cluster level configurations. Compaction related properties are one such example.  

Initially I wanted to refactor the whole thing but the patch didn't look pretty and I gave up and ended up fixing this with a simple localized fix. One thing that surprised me was that we don't allow all configs to have overrides (there is an allow-list). I remember reading in the docs that one can override any topic config at the cluster scope but I'm not able to find that documentation now (probably fixed?).

Reply:
> One thing that surprised me was that we don't allow all configs to have overrides (there is an allow-list). I remember reading in the docs that one can override any topic config at the cluster scope

I didn't quite follow that last part: do you mean the reverse "we can override any cluster config at topic scope"? Normally topic configs would "override" cluster ones (i.e., principle of highest specificity or whatever that is called).

Reply:
Oops that's not what I intended to say (came out totally opposite), thanks for catching.. I meant that not all topic-related configs can have topic-level overrides, [documentation] (https://docs.redpanda.com/docs/21.11/cluster-administration/configuration/) says the following and that gave me an impression that any topic-level configuration can have overrides.

```
Node settings in the Redpanda configuration file apply to topics if there is no topic-level setting specified through rpk or another Kafka client. This means that if you update the node configuraiton in redpanda.yaml, topic configuration will reflect those changes. However, if you configure a setting at the topic-level, that setting overrides the node-level setting and will not be affected by changes in redpanda.yaml.
```

That said, this particular note has been removed in the [latest version](https://docs.redpanda.com/docs/cluster-administration/configuration/) of the docs, so I think we are good.